third-party/qthread/qthread-1.10/src/interfaces/chapel/tasks-qthreads.c

```diff
@@ -615,13 +618,125 @@ void chpl_task_exit(void)
 #endif /* QTHREAD_MULTINODE */
 }
 
+//
+// Tasking callback support.
+//
+#define MAX_CBS_PER_EVENT 10
+
+static struct cb_info {
+    chpl_task_cb_fn_t fns[MAX_CBS_PER_EVENT];
+    chpl_task_cb_info_kind_t info_kinds[MAX_CBS_PER_EVENT];
+    int count;
+} cb_info[chpl_task_cb_num_event_kinds];
+
+int chpl_task_install_callback(chpl_task_cb_event_kind_t event_kind,
+                               chpl_task_cb_info_kind_t info_kind,
+                               chpl_task_cb_fn_t cb_fn) {
+    int i;
+
+    if (event_kind >= chpl_task_cb_num_event_kinds) {
+        errno = ERANGE;
+        return -1;
+    }
+
+    i = cb_info[event_kind].count;
+
+    if (i >= MAX_CBS_PER_EVENT) {
+        errno = ENOMEM;
+        return -1;
+    }
+
+    cb_info[event_kind].count++;
+    cb_info[event_kind].fns[i]= cb_fn;
+    cb_info[event_kind].info_kinds[i] = info_kind;
+
+    return 0;
+}
+
+int chpl_task_uninstall_callback(chpl_task_cb_event_kind_t event_kind,
+                                 chpl_task_cb_fn_t cb_fn) {
+    int i;
+    int found_i;
+
+    if (event_kind >= chpl_task_cb_num_event_kinds) {
+        errno = ERANGE;
+        return -1;
+    }
+
+    for (i = 0, found_i = -1; i < cb_info[event_kind].count; i++) {
+        if (cb_info[event_kind].fns[i] == cb_fn) {
+            found_i = i;
+            break;
+        }
+    }
+
+    if (found_i < 0) {
+        errno = ENOENT;
+        return -1;
+    }
+
+    for (i = found_i + 1; i < cb_info[event_kind].count; i++) {
+        cb_info[event_kind].fns[i - 1] =
+            cb_info[event_kind].fns[i];
+        cb_info[event_kind].info_kinds[i - 1] =
+            cb_info[event_kind].info_kinds[i];
+    }
+
+    cb_info[event_kind].count--;
+
+    return 0;
+}
+
+static inline void do_callbacks(chpl_task_cb_event_kind_t event_kind,
```

Comment:
Ordering. I'd have to move the one in fifo before its callers in order for it to be inlined (as I understand it). I'll add that to my list, though. (In each case I followed the existing style, and fifo has callees after their callers in most cases, while qthreads has them before.)